llvm/include/llvm/IR/DebugProgramInstruction.h

```diff
@@ -69,6 +69,53 @@ class DPMarker;
 class DPValue;
 class raw_ostream;
 
+/// A typed tracking MDNode reference that does not require a definition for its
+/// parameter type. Necessary to avoid including DebugInfoMetadata.h, which has
+/// a significant impact on compile times if included in this file.
+template <typename T> class DbgRecordParamRef {
```

Comment:
Is this tactic used anywhere else? If so, would it be worth putting in its own header / the TypedTrackingMDNodeRef header?

Reply:
This tactic is used for DebugLoc; that case specializes further for `DILocation`s though, such that we can't trivially share a definition - at the very least we wouldn't deduplicate very much code. In principle I think this could be available in a broader header, //but// imo it should be done at the point we need it rather than preemptively.